### ISSUE TITLE
Negative temperatures

### ISSUE THREAD

**Issue by blinkingbee** [tobyweston/temperature-machine#13]
Problem with temperatures below zero being logged as large positive numbers (e.g. 4085.3 °C). This is also reflected in the graphs auto-scaling to 4000 °C. 

When I read the sensor temperature using a python script with an expression such as  `echo `/usr/local/bin/get_temp.py | head -1 | awk -F. '{print $1"."int(($2/100)+0.5)}'`` this returns the correct negative temperature. 

Could this be a problem with how Temperature-Machine reads the sensor and handles signing?

![negative temp example](https://user-images.githubusercontent.com/29305609/26975680-c495ee4a-4d18-11e7-9ed5-2235c11d71b5.png)

[get_temp.py.zip](https://github.com/tobyweston/temperature-machine/files/1064017/get_temp.py.zip)

**Comment by tobyweston** [comment]
I can reproduce it with the help of an ice pack! 
```
e1 ff 4b 46 7f ff 0c 10 fe : crc=fe YES
e1 ff 4b 46 7f ff 0c 10 fe t=-1937
```

![screen shot 2017-06-09 at 20 24 59](https://user-images.githubusercontent.com/49573/26991365-ec70a782-4d51-11e7-863f-d2d6e0cd06a6.png)

I think there's a problem in `Parser.scala`, I was trying to be all fancy and calculate the temperature using the raw byte values rather than use the scratch file pre-calculated value (```t=-1.2```). It has the advantage of being more precise. See http://temperature-machine.com/docs/ds18b20_sensor.html 

I may have been a bit naive with my above calculation. See [the spec sheet](https://cdn-shop.adafruit.com/datasheets/DS18B20.pdf): I think i'll be better off doing the proper two's complement on the bytes.

**Comment by blinkingbee** [comment]
Samples taken at 3 different temperatures with icepack from my freezer:
[w1_slave samples.zip](https://github.com/tobyweston/temperature-machine/files/1065257/w1_slave.samples.zip)

It would be very useful if you could use the pre-calculated value from the scratchpad and thus exchange some precision for valid temperatures below zero.

**Comment by blinkingbee** [comment]
Posted some examples for both +ve and -ve temperatures

**Comment by blinkingbee** [comment]
I have updated Parser.scala and ParserTest.scala in both client and server but temperatures below zero still display as large positive numbers. Is there anything else I need to do?

**Comment by tobyweston** [comment]
I haven't deployed myself yet so am still testing. I'll prepare a 'package' once done and close the issue here.

How are you deploying (see http://temperature-machine.com/docs/getting_started.html#setup-the-software).

If you're building from source, you can test by following [these instructions](http://temperature-machine.com/docs/getting_started/build_from_source.html) having first done the following;

```
$ git pull
$ git checkout issue_13
```

**Comment by blinkingbee** [comment]
Thanks, maybe I jumped the gun here. I originally deployed using the pre-installed image. If I build from source does this just overwrite the original temperature-machine?

**Comment by tobyweston** [comment]
Yep, on the pre-installed image under `~/code/temperature-machine`. Just run the above and rebuild the `.jar` file (steps 2. from [these instructions](http://temperature-machine.com/docs/getting_started/build_from_source.html)).

So basically,

```
cd ~/code/temperature-machine
git pull
git checkout issue_13
sbt -J-Xmx512m -J-Xms512m assembly
```
(You can drop/alter the `-J` settings if you're not running on a Pi Zero)

Then restart.

__Note__ that this will checkout the _branch_ so once it's merged back into `master`, you'll have to reverse the procedure (i.e, `git checkout master` then `sbt assembly`). I'll merge back after testing and building a new image etc (which will prob take a little while)

Let me know how you get on.

Oh, and I tend to take down the server / hub first to upgrade, then take down each child node. To be super safe, you can take them all down and update, server first.

**Comment by tobyweston** [comment]
Just merged bf10a84 to master so ignore the `git checkout issue_13` step...

**Comment by blinkingbee** [comment]
I have rebuilt both master and slave as described above but am still getting large positive numbers when temperature goes below zero. Separately monitoring the sensor scratch file pre-calculated value shows the correct negative temperature.

**Comment by tobyweston** [comment]
and the `sbt assembly` step worked for you? Sometimes it crashes on the Pi because of the limited resources and you have to re-run. Check the timestamp of the `temperature-machine.jar` file. Something like `find . -name "*.jar"` from the temperature-machine fold should locate it. Make sure it's been updated.

**Comment by blinkingbee** [comment]
`-rw-r--r-- 1 pi pi 28515459 Apr  2 14:09 temperature-machine-2.0.jar` It looks like it's not been updated so I'll try a rerun

**Comment by tobyweston** [comment]
Cool, when you run `sbt assembly` (on the Pi zero in particular), it can fail more or less silently. Re-running usually works (but __don't__ `sbt clean` first).

**Comment by blinkingbee** [comment]
Mostly success. My two client units built successfully and the one I've tested displays negative temperatures correctly. However the server returns errors and won't build. I have attached the console output for your comment. I may have to start over with the server if I can't solve it.

```
$ sbt -J-Xmx512m -J-Xms512m assembly
...
[error] Not a valid command: J-Xms512m
[error] Not a valid key: J-Xms512m
[error] J-Xms512m
[error]
```

[attempted build.txt.zip](https://github.com/tobyweston/temperature-machine/files/1071035/attempted.build.txt.zip)
